Kernel/Arch/aarch64/PCI/Controller/BCM2712HostController.cpp

```diff
@@ -0,0 +1,263 @@
+// The config space is accessible through a single 4K window that can be mapped to any bus/device/function.
+// The b/d/f can be configured by writing an ECAM offset to Registers::config_space_window_address.
+
+struct Registers {
+    u8 unknown1[0x400c];
+
+    // Bus base [31:0] @ [31:0]
+    u32 bus_window_base_low; // 0x400c
+
+    // Bus base [63:32] @ [31:0]
+    u32 bus_window_base_high; // 0x4010
+
+    u8 unknown2[0x4064 - (0x4010 + 4)];
+
+    u32 control; // 0x4064
+    static constexpr u32 CONTROL_PERST_N = 1 << 2;
+
+    u32 state; // 0x4068
+    static constexpr u32 STATE_LINK_UP = 0b11 << 4;
+
+    u8 unknown3[0x4070 - (0x4068 + 4)];
+
+    // The host controller doesn't seem to support 16-bit accesses, so the base and limit are grouped into one field.
+    // CPU limit [31:20] @ [31:20]
+    // CPU base [31:20] @ [15:4]
+    u32 cpu_window_base_low_and_cpu_window_limit_low; // 0x4070
+
+    u8 unknown4[0x4080 - (0x4070 + 4)];
+
+    // CPU base [39:32] @ [7:0]
+    u32 cpu_window_base_high; // 0x4080
+
+    // CPU limit [39:32] @ [7:0]
+    u32 cpu_window_limit_high; // 0x4084
+
+    u8 unknown5[0x8000 - (0x4084 + 4)];
+
+    u8 config_space_window[0x1000];  // 0x8000
+    u32 config_space_window_address; // 0x9000
+
+    u8 unknown6[0x9310 - (0x9000 + 4)]; // 0x9000
+};
+static_assert(AssertSize<Registers, 0x9310>());
+
+class BCM2712HostController : public HostController {
+public:
+    static ErrorOr<NonnullOwnPtr<BCM2712HostController>> create(DeviceTree::Device const& device);
+
+private:
+    ErrorOr<VirtualAddress> get_config_window(BusNumber, DeviceNumber, FunctionNumber);
+
+    virtual void write8_field_locked(BusNumber, DeviceNumber, FunctionNumber, u32 field, u8 value) override;
+    virtual void write16_field_locked(BusNumber, DeviceNumber, FunctionNumber, u32 field, u16 value) override;
+    virtual void write32_field_locked(BusNumber, DeviceNumber, FunctionNumber, u32 field, u32 value) override;
+
+    virtual u8 read8_field_locked(BusNumber, DeviceNumber, FunctionNumber, u32 field) override;
+    virtual u16 read16_field_locked(BusNumber, DeviceNumber, FunctionNumber, u32 field) override;
+    virtual u32 read32_field_locked(BusNumber, DeviceNumber, FunctionNumber, u32 field) override;
+
+    explicit BCM2712HostController(PCI::Domain const&, Memory::TypedMapping<Registers volatile>);
+
+    Memory::TypedMapping<Registers volatile> m_registers;
+};
+
+ErrorOr<NonnullOwnPtr<BCM2712HostController>> BCM2712HostController::create(DeviceTree::Device const& device)
+{
+    auto domain = TRY(determine_pci_domain_for_devicetree_node(device.node(), device.node_name()));
+    auto registers_resource = TRY(device.get_resource(0));
+
+    if (registers_resource.size < sizeof(Registers))
+        return ERANGE;
+
+    auto registers = TRY(Memory::map_typed_writable<Registers volatile>(registers_resource.paddr));
+
+    // Deassert PERST# to initialize the link.
+    registers->control |= Registers::CONTROL_PERST_N;
```

Comment:
OK, `|=` into a `volatile` seems as iffy as bit fields...

Reply:
Why is `|=` worse than bit fields?

Bit fields imho just have too many footguns like GCC optimizing bit fields accesses into smaller width accesses even with `volatile`. But I guess we could set `-fstrict-volatile-bitfields` for the entire kernel.

Reply:
Not worse, but also iffy, IIRC the standard should actually forbid this, as `|=` is not an atomic operation

Reply:
Setting a bit field is also a read-modify-write operation and I would argue a even more dubious one. So I think it's a bit weird that the standard allows volatile bit field writes.
Does C++ even specify the access width used for volatile bit field writes?

I can rewrite it has `registers->control = registers->control | Registers::CONTROL_PERST_N;` if you think that's better.

Reply:
you can leave it be, I mainly remember getting a compiler error in the past about some modifying assignments, so beware Ig